[Qt] RPC-Console: support nested commands and simple value queries

jonasschnelli opened the pull request:
**this is currently limited to the QT RPC Console, but I'm happy to extend/refactor this to make it useable in bitcoin-cli**

Commands can be executed with bracket syntax, example: `getwalletinfo()`.
Commands can be nested, example: `sendtoaddress(getnewaddress(), 10)`.
Simple queries are possible: `listunspent()[0][txid]`
Object values are accessed with a non-quoted string, example: `[txid]`.

Fully backward compatible.
`generate 101` is identical to `generate(101)`
Result value queries indicated with `[]` require the new brackets syntax.
Comma as argument separator is now also possible: `sendtoaddress,<address>,<amount>`
Space as argument separator works also with the bracket syntax, example: `sendtoaddress(getnewaddress() 10)`

No dept limitation, complex commands are possible:
`decoderawtransaction(getrawtransaction(getblock(getbestblockhash())[tx][0]))[vout][0][value]`

<img width="1336" alt="bildschirmfoto 2016-04-01 um 17 13 08" src="https://cloud.githubusercontent.com/assets/178464/14211198/07825dca-f82d-11e5-90c4-1a44b715b2f3.png">

luke-jr commented:
Isn't this going a bit overboard for debugging tools? (OTOH, it's only about 150 LOC...)

MarcoFalke commented:
I think this is useful. An alternative proposed was to use variables `a = getnewaddress` `sendtoaddress a 10`, but this pull is fine as well.

jonasschnelli commented:
> Isn't this going a bit overboard for debugging tools? (OTOH, it's only about 150 LOC...)

It is a "luxury extension", right. But given the time some of us have spent in the console repeating and copy-pasting commands out- and input, I think it worth taking this in.
Also, I don't see critical risks for this.

IMO we should also extend bitcoin-cli to support nested commands. It simply increases productivity with that tool.

laanwj commented:
I like this concept. I initially had @luke-jr 's concern as well. But only a bit of code added, and it's well-contained.

It's not just luxury: it's useful for cases like #7599 where someone wants to insert the output of a previous command into a new one, but it's too long for copy pasting.

> IMO we should also extend bitcoin-cli to support nested commands. It simply increases productivity with that tool.

Not sure there. `bitcoin-cli` is simply an entry point and you can use whatever scripting you want in the shell that calls it.

For the GUI debug console, which is essentially it's own `bash`, this makes more sense.

I'm all for a fancy ncurses-based interactive client, but that should not be `bitcoin-cli`.

paveljanik commented:
On the other hand, many advices read 'Run getsomething' and so. This will bring another "fork" - you have to also add that you have to run this in Debug console or via `bitcoin-cli`, because the syntax will "fork".

Concept ACK (I'd also like to see this in `bitcoin-cli`)

luke-jr commented:
Almost tempting to make it server-side, if we're using long output-inputs... but this seems fine (Concept ACK) as-is; further improvement can wait for another PR.

sipa commented:
Would it be possible to abstract out this functionality in a separate commit (including the existing RPC parsing logic from the Qt console) and move it to rpc/server.cpp, as an actual RPC call that just takes a string argument with a command to parse?

That would make it both more usable (by exposing it as RPC, bitcoin-cli and other tools can use it too), and more testable (we can have RPC tests for it), without complicating the change much.

jonasschnelli commented:
I have though about that but I wasn't sure if we should delegate the parsing/executing of nested command to the server. This PR would do the parsing "client side". We could also factor out the parsing and use it client-side (Qt / bitcoin-cli).

But I agree, it could be useful server-side.

sipa commented:
Yes, I agree having it client side is just as useful. My main reason for suggesting abstracting it out it because I don't think it's very hard, and would make the parsing logic much easier to test.

UniQredit commented:
This would save a hell lot of time and copy/pasting

jonasschnelli commented:
* Finally rebased this great PR and refactored the parsing logic into `rpc/server.cpp`
* Added some unit tests

This could now be simply extended to the RPC server, although, nested commands could be resource and time hungry.

MarcoFalke commented on src/test/rpc_tests.cpp:
```diff
@@ -332,4 +332,49 @@ BOOST_AUTO_TEST_CASE(rpc_convert_values_generatetoaddress)
+    BOOST_CHECK_NO_THROW(RPCExecuteCommandLine(result, "getblockchaininfo()"));
+    BOOST_CHECK_EQUAL(result.substr(0,1), "{");
+
+    BOOST_CHECK_NO_THROW(RPCExecuteCommandLine(result, "getblockchaininfo")); //whitespace at the end will be tolerated
```
There is no whitespace at the end?

MarcoFalke commented on src/test/rpc_tests.cpp:
```diff
@@ -332,4 +332,49 @@ BOOST_AUTO_TEST_CASE(rpc_convert_values_generatetoaddress)
+
+    BOOST_CHECK_THROW(RPCExecuteCommandLine(result, "getblockchaininfo() .\n"), runtime_error); //invalid syntax
+    BOOST_CHECK_THROW(RPCExecuteCommandLine(result, "getblockchaininfo() getblockchaininfo()"), runtime_error); //invalid syntax
+    BOOST_CHECK_NO_THROW(RPCExecuteCommandLine(result, "getblockchaininfo(")); //tollerate non closing brackets if we have no arguments
```
`:%s/toller/toler/g`

jonasschnelli commented:
Fixed nits. Had to add `client.cpp` to `libbitcoin_server_a`. Should that be a problem @theuni?

laanwj commented:
> Would it be possible to abstract out this functionality in a separate commit (including the existing RPC parsing logic from the Qt console) and move it to rpc/server.cpp,

Sorry to be contrary, but IMO, functionality related to parsing and not dispatching should be in `rpc/client.cpp` instead of `rpc/server.cpp`. Note that bitcoin-cli links the client library, not the server one.

laanwj commented on src/Makefile.am:
```diff
@@ -181,6 +181,7 @@ libbitcoin_server_a_SOURCES = \
   pow.cpp \
   rest.cpp \
   rpc/blockchain.cpp \
+  rpc/client.cpp \
```
client.cpp is part of `libbitcoin_cli`, which is "cli: shared between bitcoin-cli and bitcoin-qt" if you need access to that then link that library. Don't include the compilation unit in two libraries.

jonasschnelli commented:
Having it in `rpc/client.cpp` instead of `server.cpp` would be good I guess. The only restriction would then be, that we cannot allow (later) server side nested commands (which would probably perform slightly faster). But I'm not sure if we really want server side nested commands with the current locking behavior.

laanwj commented:
Server-side nested commands are not part of the JSON-RPC standard. It is an interesting thought but that would be a completely different proposal, and I don't think it would share any code with this. I'd imagine it would work something akin to batching (but w/ nested structures), not by parsing/formatting expression strings.

theuni commented:
Agreed with keeping parsing client-side. Let's not tangle up the dependencies.

I really like this idea btw.

jonasschnelli committed:
[Qt] RPC-Console: support nested commands and simple value queries

Commands can be executed with bracket syntax, example: `getwalletinfo()`.
Commands can be nested, example: `sendtoaddress(getnewaddress(), 10)`.
Simple queries are possible: `listunspent()[0][txid]`
Object values are accessed with a non-quoted string, example: [txid].

Fully backward compatible.
`generate 101` is identical to `generate(101)`
Result value queries indicated with `[]` require the new brackets syntax.
Comma as argument separator is now also possible: `sendtoaddress,<address>,<amount>`
Space as argument separator works also with the bracket syntax, example: `sendtoaddress(getnewaddress() 10)

No dept limitation, complex commands are possible:
`decoderawtransaction(getrawtransaction(getblock(getbestblockhash())[tx][0]))[vout][0][value]`

jonasschnelli commented:
Removed all changes from the core classes.
It's now a GUI only change.
Added Qt unit tests for the nested commands.

MarcoFalke commented:
qt-test fail on travis, apparently.

jonasschnelli commented:
Fixed the travis Qt-Test issue.
This PR is looking for reviewers.

UdjinM6 commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,211 @@ class QtRPCTimerInterface: public RPCTimerInterface
+                                {
+                                    // if we have a value query, query arrays with index and objects with a string key
+                                    UniValue subelement;
+                                    if (curarg.size() && lastResult.isArray())
```
`curarg.size()` should already be ok since it's checked in line 170

UdjinM6 commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,211 @@ class QtRPCTimerInterface: public RPCTimerInterface
+                                                throw std::runtime_error("Invalid result query");
+                                        subelement = lastResult[atoi(curarg.c_str())];
+                                    }
+                                    else if (curarg.size() && lastResult.isObject())
```
same here for `curarg.size()`

UdjinM6 commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,211 @@ class QtRPCTimerInterface: public RPCTimerInterface
+                    default:
+                        if (state == STATE_COMMAND_EXECUTED_INNER)
+                        {
+                            if (ch == ']')
```
nit: since both branches for that `if` end with `break`, nesting level here can be reduced by smth like:
```
                            if (ch != ']')
                            {
                                // append char to the current argument (which is also used for the query command)
                                curarg += ch;
                                break;
                            }
```
and then all the code in the scope below but up one level of nesting.

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,211 @@ class QtRPCTimerInterface: public RPCTimerInterface
+                                strResult = curarg;
+                        }
+                        curarg.clear();
+                        // assume easting space state
```
nit: typo

UdjinM6 commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,211 @@ class QtRPCTimerInterface: public RPCTimerInterface
-
-        Q_EMIT reply(RPCConsole::CMD_REPLY, QString::fromStdString(strPrint));
+        std::string result;
+        std::string executableCommand = command.toStdString() + "\n";
```
nit: maybe change names to `strResult` and `strExecutableCommand`? Same for strings in rpcnestedtests.cpp

MarcoFalke commented on src/qt/test/rpcnestedtests.cpp:
```diff
@@ -0,0 +1,87 @@
+    const CChainParams& chainparams = Params();
+    RegisterAllCoreRPCCommands(tableRPC);
+    ClearDatadirCache();
+    std::string path = QDir::tempPath().toStdString() + "/" + strprintf("test_bitcoin_%lu_%i", (unsigned long)GetTime(), (int)(GetRand(100000)));
```
Any reason for this name?

MarcoFalke replied:
Also I feel like this should be cleaned up on exit?

jonasschnelli commented:
Fixed nits, added cleanup of Qt test data.

dcousens commented:
concept ACK

laanwj commented:
utACK.

I do think this needs documentation. Not necessarily in this pull, but currently the debug console help consists of two lines "Use up and down arrows to navigate history, and Ctrl-L to clear screen. Type help for an overview of available commands.". 

Maybe add a debug-console-only command like `help` that shows how to use nested commands and potentially other advanced tricks, and add 'for more information on using this console type XXX'.

jonasschnelli commented:
> I do think this needs documentation.  [...]

Good point. I try something. Maybe not in this PR.
Could – maybe – be combined with this https://github.com/bitcoin/bitcoin/issues/8544#issuecomment-240970925

laanwj commented:
`getwalletinfo()["walletversion"]` doesn't work - can't it index into objects?

jonasschnelli commented:
@laanwj: I guess you need to use `getwalletinfo()[walletversion]` (without double-quotes).

laanwj commented:
@jonasschnelli awesome, that works. So that's why we need documentation :)

jonasschnelli commented:
@laanwj: Agree on the documentation. The dropped `"`(double quotes) for an index access is quite uncommon, but can make sense because all our JSON properties are pure ASCII without whitespace.

Allowing the double-quotes (ignore them while parsing) could be a useful addition.

laanwj commented:
For testing this it's useful to add an `echo` command that simply returns what is passed to it:
```patch
diff --git a/src/rpc/misc.cpp b/src/rpc/misc.cpp
index 5afcf63..9272c53 100644
--- a/src/rpc/misc.cpp
+++ b/src/rpc/misc.cpp
@@ -417,6 +417,17 @@ UniValue signmessagewithprivkey(const UniValue& params, bool fHelp)
     return EncodeBase64(&vchSig[0], vchSig.size());
 }
 
+UniValue echo(const UniValue& params, bool fHelp)
+{
+    if (fHelp)
+        throw runtime_error(
+            "echo \"message\" ...\n"
+            "\nSimply echo back the input arguments\n"
+        );
+
+    return params;
+}
+
 UniValue setmocktime(const UniValue& params, bool fHelp)
 {
     if (fHelp || params.size() != 1)
@@ -454,6 +465,7 @@ static const CRPCCommand commands[] =
 { //  category              name                      actor (function)         okSafeMode
   //  --------------------- ------------------------  -----------------------  ----------
     { "control",            "getinfo",                &getinfo,                true  }, /* uses wallet if enabled */
+    { "control",            "echo",                   &echo,                   true  },
     { "util",               "validateaddress",        &validateaddress,        true  }, /* uses wallet if enabled */
     { "util",               "createmultisig",         &createmultisig,         true  },
     { "util",               "verifymessage",          &verifymessage,          true  },
```

> Allowing the double-quotes (ignore them while parsing) could be a useful addition.

Yes, would be useful to add, maybe in a later pull. Somehow I keep typing [] accidentally, and being surprised I get 'null'.

laanwj commented:
ACK 2ca6b9d

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,208 @@ class QtRPCTimerInterface: public RPCTimerInterface
  *   - Within single quotes, no escaping is possible and no special interpretation takes place
  *
- * @param[out]   args        Parsed arguments will be appended to this list
+ * @param[out]   result      stringified Result from the executed command(chain)
```
Nit: `strResult`

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -113,117 +113,208 @@ class QtRPCTimerInterface: public RPCTimerInterface
+        std::string executableCommand = command.toStdString() + "\n";
+        if(!RPCConsole::RPCExecuteCommandLine(result, executableCommand))
+        {
+            Q_EMIT reply(RPCConsole::CMD_ERROR, QString("Parse error: unbalanced ' or \""));
```
Nit: `[` can also be unbalanced

MarcoFalke commented on src/qt/test/rpcnestedtests.cpp:
```diff
@@ -0,0 +1,93 @@
+    QVERIFY_EXCEPTION_THROWN(RPCConsole::RPCExecuteCommandLine(result, "getblockchaininfo() .\n"), std::runtime_error); //invalid syntax
+    QVERIFY_EXCEPTION_THROWN(RPCConsole::RPCExecuteCommandLine(result, "getblockchaininfo() getblockchaininfo()"), std::runtime_error); //invalid syntax
+    (RPCConsole::RPCExecuteCommandLine(result, "getblockchaininfo(")); //tolerate non closing brackets if we have no arguments
+    (RPCConsole::RPCExecuteCommandLine(result, "getblockchaininfo()()()")); //tolerate non command brackts
```
Nits: Can be moved out of the guard, typo brackts, no wrapping brackets needed.
